File: Python/49_Group_Anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = {} # make unordered dict
        for s in strs:
            char_counts = [0] * 26
            for c in s:
                char_counts[ord(c) - 97] += 1
            
            key = tuple(char_counts)

            if key in res:
                res[key].append(s)
            else:
                res[key] = [s]
        
        return list(res.values())

File: Cpp/49_Group_Anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> res;

        for (string& s : strs) {
            vector<int> char_counts(26, 0);
            for (char c : s) {
                char_counts[c - 'a']++;
            }

            // convert counts into a string key
            string key;
            for (int count : char_counts) {
                key += to_string(count) + "#"; 
            }

            res[key].push_back(s);
        }

        vector<vector<string>> ans;
        for (auto& pair : res) {
            ans.push_back(move(pair.second));
        }

        return ans;
    }
};